PyInstaller/depend/analysis.py

```diff
@@ -156,7 +124,85 @@ def run_script(self, pathname, caller=None):
                 caller = self._top_script_node
             return super(PyiModuleGraph, self).run_script(pathname, caller=caller)
 
-    def _safe_import_module(self, module_basename, module_name, parent_package):
+    def _is_excluded_import(self, module_name, imported_by):
+        """
+        If a module is in the excluded imports dict then either keep it there or
+        remove it from the dict.
+
+        Conditions for keeping excluded module in the dict:
+        - Any parent package of 'module_name' is in 'excluded_imports' and
+          - 'imported_by' is a subpackage of 'module_name.
+          - 'imported_by' is a subpackage of any item from the 'excluded_imports'
+            dict value.
+
+        :param module_name: Module name to check if keeping in excluded modules.
+        :param imported_by: Module name that imports module 'module_name'.
+        :return: True if the module should be really excluded or False otherwise.
+        """
+        # TODO check if it is necessary remember references to 'imported_by' when exludedimports are later included!
+        #      If there are later reported any import errors with missing modules then this is probably the case.
+        # Check 'module_name' and all its parent packages.
+        parent_packages = [module_name] + module_parent_packages(module_name)
+        for module in parent_packages:
+            # Check if there is any excluded import for the module.
+            if module in self.excluded_imports:
+                # TODO verify if 'imported_by' being subpackage of excludedimports needs checking.
+                # 'imported_by' is a subpackage of 'module', keep the excluded
+                # import and just end the function.
+                #if imported_by.startswith(module):
+                    # Add submodules of excluded modules to the exclude list and
+                    # exclude it for the same modules as the parent package.
+                    #self.excluded_imports[module_name] = self.excluded_imports[module]
+                    #return True
+                # Modules for which the excluded import should be kept.
+                not_allowed_modules = self.excluded_imports[module]
+                # 'imported_by' is not allowed or is subpackage of any not
+                # allowed module.
+                for not_allowed in not_allowed_modules:
+                    if imported_by.startswith(not_allowed):
```

Comment:
`if imported_by.startswith(not_allowed) + '.' or imported_by == not_allowed`.

Otherwise `foo` would exclude `foobar`, too.

It could be worth storing the `module_parent_packages(not_allowed)` in `self.excluded_imports`